Fix catalog configuration file search.

levkohimins opened the pull request:
## Description

Fixed the issue with finding a configuration file when running an application in the same directory as the configuration file.

Fixes #2861.

brikis98 reviewed (commented):
It looks like this would fix finding a `terragrunt.hcl` in the current working dir, but will this fix the issue where that file includes something like `find_in_parent_dir()`?

levkohimins commented:
> It looks like this would fix finding a `terragrunt.hcl` in the current working dir, but will this fix the issue where that file includes something like `find_in_parent_dir()`?

Yes, it will. To be sure, I've added your example as a test case.

brikis98 commented:
> > It looks like this would fix finding a `terragrunt.hcl` in the current working dir, but will this fix the issue where that file includes something like `find_in_parent_dir()`?
> 
> Yes, it will. To be sure, I've added your example as a test case.

Ah, OK... And how is it working? E.g.,, if the root `terragrunt.hcl` has a `find_in_parent_folders("common.hcl")`, and you run `terragrunt catalog` from the root folder... How do you not get an error about not finding `common.hcl` in parent folders?

levkohimins commented:
> > > It looks like this would fix finding a `terragrunt.hcl` in the current working dir, but will this fix the issue where that file includes something like `find_in_parent_dir()`?
> > 
> > 
> > Yes, it will. To be sure, I've added your example as a test case.
> 
> Ah, OK... And how is it working? E.g.,, if the root `terragrunt.hcl` has a `find_in_parent_folders("common.hcl")`, and you run `terragrunt catalog` from the root folder... How do you not get an error about not finding `common.hcl` in parent folders?

I got this error when running from the parent directory :)
I added a test case to check this error and complicated the config.

brikis98 commented:
> > > > It looks like this would fix finding a `terragrunt.hcl` in the current working dir, but will this fix the issue where that file includes something like `find_in_parent_dir()`?
> > > 
> > > 
> > > Yes, it will. To be sure, I've added your example as a test case.
> > 
> > 
> > Ah, OK... And how is it working? E.g.,, if the root `terragrunt.hcl` has a `find_in_parent_folders("common.hcl")`, and you run `terragrunt catalog` from the root folder... How do you not get an error about not finding `common.hcl` in parent folders?
> 
> I got this error when running from the parent directory :) I added a test case to check this error and complicated the config. It's working as expected.

Not sure I follow... We _don't_ want this to be an error. `terragrunt catalog` should work, whether you run it from the root, a subfolder with a `terragrunt.hcl`, or a subfolder with no `terragrunt.hcl`.

levkohimins commented:
> > > > > It looks like this would fix finding a `terragrunt.hcl` in the current working dir, but will this fix the issue where that file includes something like `find_in_parent_dir()`?
> > > > 
> > > > 
> > > > Yes, it will. To be sure, I've added your example as a test case.
> > > 
> > > 
> > > Ah, OK... And how is it working? E.g.,, if the root `terragrunt.hcl` has a `find_in_parent_folders("common.hcl")`, and you run `terragrunt catalog` from the root folder... How do you not get an error about not finding `common.hcl` in parent folders?
> > 
> > 
> > I got this error when running from the parent directory :) I added a test case to check this error and complicated the config. It's working as expected.
> 
> Not sure I follow... We _don't_ want this to be an error. `terragrunt catalog` should work, whether you run it from the root, a subfolder with a `terragrunt.hcl`, or a subfolder with no `terragrunt.hcl`.

Oh, sorry, I didn't read your idea carefully in the issue. I really liked it and will implement it.

levkohimins commented:
> > > > > > It looks like this would fix finding a `terragrunt.hcl` in the current working dir, but will this fix the issue where that file includes something like `find_in_parent_dir()`?
> > > > > 
> > > > > 
> > > > > Yes, it will. To be sure, I've added your example as a test case.
> > > > 
> > > > 
> > > > Ah, OK... And how is it working? E.g.,, if the root `terragrunt.hcl` has a `find_in_parent_folders("common.hcl")`, and you run `terragrunt catalog` from the root folder... How do you not get an error about not finding `common.hcl` in parent folders?
> > > 
> > > 
> > > I got this error when running from the parent directory :) I added a test case to check this error and complicated the config. It's working as expected.
> > 
> > 
> > Not sure I follow... We _don't_ want this to be an error. `terragrunt catalog` should work, whether you run it from the root, a subfolder with a `terragrunt.hcl`, or a subfolder with no `terragrunt.hcl`.
> 
> Oh, sorry, I didn't read your idea carefully in the issue. I really liked it and will implement it.

@brikis98, I has implemented your idea (you are a genius).

brikis98 commented on config/catalog.go:
```diff
@@ -27,3 +38,40 @@ func (config *CatalogConfig) normalize(cofnigPath string) {
 	}
 }
+
+func ReadCatalogConfig(opts *options.TerragruntOptions) (*TerragruntConfig, error) {
```
NIT: do you mind adding a comment here explaining what this method does and _why_?

Namely:

- We want users to be able to browse to any folder in an `infra-live` repo, run `terragrunt catalog` (with no URL) arg, and we automatically find the "closest" terragrunt.hcl, and load the catalogs from its `catalog` config.
- So this method searches the current working directory and its parent folders for a `terragrunt.hcl`.
- We then need to parse this file. Our normal parsing won't always work, as some `terragrunt.hcl` files are meant to be used from an `include` and/or they might use `find_in_parent_folders`, so they only work from certain child folders.
- So if we find a `terragrunt.hcl` that already has an `include`, we parse that as-is; if we find a `terragrunt.hcl` without `include`, we generate a stub child `terragrunt.hcl` in memory with an `include` to pull in the one we found.

levkohimins replied:
Of course, thanks for the example text. I will try not to miss important explanations of functions in the future.

brikis98 reviewed (approved):
LGTM!
